axelor-sale/src/main/java/com/axelor/apps/sale/service/configurator/ConfiguratorServiceImpl.java

```diff
@@ -177,7 +244,147 @@ public void generateProduct(
+                        && fieldName.equals(formula.getMetaJsonField().getName())) {
+                      putFieldValueInMap(
+                          fieldName,
+                          jsonIndicators.get(fullName),
+                          attrName,
+                          formula.getMetaJsonField(),
+                          attrValueMap);
+                      keysToRemove.add(fullName);
+                    }
+                  });
+            });
+
+    jsonIndicators.entrySet().removeIf(entry -> keysToRemove.contains(entry.getKey()));
+    return attrValueMap;
+  }
+
+  private void putFieldValueInMap(
+      String nameField,
+      Object object,
+      String attrName,
+      MetaJsonField metaJsonField,
+      Map<String, Map<String, Object>> attrValueMap) {
+
+    if (!attrValueMap.containsKey(attrName)) {
+      attrValueMap.put(attrName, new HashMap<>());
+    }
+    Entry<String, Object> entry = adaptType(nameField, object, metaJsonField);
+    attrValueMap.get(attrName).put(entry.getKey(), entry.getValue());
+  }
+
+  /**
+   * Private method that adapt type of object depending on his type.
+   *
+   * @param nameField
+   * @param object
+   * @param metaJsonField
+   * @return
+   */
+  private Map.Entry<String, Object> adaptType(
+      String nameField, Object object, MetaJsonField metaJsonField) {
+    try {
+
+      if (object instanceof Temporal) {
+        return new AbstractMap.SimpleEntry<>(nameField, object.toString());
+      }
+
+      String wantedType = MetaTool.jsonTypeToType(metaJsonField.getType());
+      // Case of many to one object
+      if ("ManyToOne".equals(wantedType) || "Custom-ManyToOne".equals(wantedType)) {
+        // The cast should not be a problem, since at this point it must be a Model
+        final Map<String, Long> manyToOneObject = modelToJson((Model) object);
+
+        return new AbstractMap.SimpleEntry<>(nameField, manyToOneObject);
+      } else if ("OneToMany".equals(wantedType)
+          || "Custom-OneToMany".equals(wantedType)
+          || "ManyToMany".equals(wantedType)
+          || "Custom-ManyToMany".equals(wantedType)) {
+
+        @SuppressWarnings("unchecked")
+        List<Model> listModels = (List<Model>) object;
+        List<Map<String, Long>> mappedList =
+            listModels.stream().map(model -> modelToJson(model)).collect(Collectors.toList());
+
+        return new AbstractMap.SimpleEntry<>(nameField, mappedList);
+      }
+    } catch (AxelorException | IllegalArgumentException | SecurityException e) {
+
+      TraceBackService.trace(e);
+    }
+    return new AbstractMap.SimpleEntry<>(nameField, object);
+  }
+  /**
+   * This method map a model in json that need to be used in a OneToMany type
+   *
+   * @param model
+   * @return
+   */
+  protected Map<String, Long> modelToJson(Model model) {
+    final Map<String, Long> manyToOneObject = new HashMap<>();
+    manyToOneObject.put("id", model.getId());
+    return manyToOneObject;
   }
```

Comment:
Maybe all these methods should be moved in a new separate service, what do you think ?

Reply:
Yes, I was actually thinking the same, but I wasn't so sure about this choice. I'll make a new service